File: subpackage/computer-use-mcp/core/src/capability.rs
use serde::Serialize;

#[derive(Serialize)]
pub struct Capabilities {
    pub has_display: bool,
    pub platform: String,
    pub display_server: Option<String>,
    pub display_resolution: Option<(u32, u32)>,
    pub scale_factor: f64,
    pub accessibility_granted: bool,
    pub is_admin: bool,
    pub issues: Vec<String>,
}

pub fn detect_capabilities() -> Capabilities {
    let platform = std::env::consts::OS.to_string();
    let (has_display, display_server) = detect_display_server();
    let display_resolution = if has_display {
        detect_resolution().ok()
    } else {
        None
    };
    let scale_factor = display_resolution
        .map(|(w, _)| if w > 2000 { w as f64 / 1400.0 } else { 1.0 })
        .unwrap_or(1.0);
    let accessibility_granted = check_accessibility_permission();
    let is_admin = check_admin_status();

    let mut issues = Vec::new();

    if !has_display {
        issues.push("No display server detected. Computer use tools unavailable.".into());
    }

    if has_display && !accessibility_granted && platform == "macos" {
        issues.push(
            "Accessibility permission required. Grant in: \
             System Preferences → Privacy & Security → Accessibility → \
             Enable AgentBoster"
                .into(),
        );
    }

    Capabilities {
        has_display,
        platform,
        display_server,
        display_resolution,
        scale_factor,
        accessibility_granted,
        is_admin,
        issues,
    }
}

fn detect_display_server() -> (bool, Option<String>) {
    #[cfg(target_os = "macos")]
    {
        (true, Some("quartz".into()))
    }

    #[cfg(target_os = "windows")]
    {
        (true, Some("win32".into()))
    }

    #[cfg(target_os = "linux")]
    {
        if std::env::var("WAYLAND_DISPLAY").is_ok() {
            (true, Some("wayland".into()))
        } else if std::env::var("DISPLAY").is_ok() {
            (true, Some("x11".into()))
        } else {
            (false, None)
        }
    }

    #[cfg(not(any(target_os = "macos", target_os = "windows", target_os = "linux")))]
    {
        (false, None)
    }
}

fn detect_resolution() -> Result<(u32, u32), Box<dyn std::error::Error>> {
    let monitors = xcap::Monitor::all()?;
    let m = monitors.first().ok_or("No monitors")?;
    Ok((m.width(), m.height()))
}

fn check_accessibility_permission() -> bool {
    #[cfg(target_os = "macos")]
    {
        unsafe { accessibility_sys::AXIsProcessTrusted() }
    }

    #[cfg(not(target_os = "macos"))]
    {
        true
    }
}

fn check_admin_status() -> bool {
    #[cfg(unix)]
    {
        unsafe { libc::geteuid() == 0 }
    }

    #[cfg(not(unix))]
    {
        false
    }
}

File: subpackage/computer-use-mcp/pkg/capability/capability.go
package capability

import (
	"os"
	"os/exec"
	"runtime"

	"github.com/nekisekaimoe/agentboster/subpackages/computer-use-mcp/pkg/screenshot"
)

// hasCmd reports whether an executable is on PATH. Used to detect the optional
// Wayland helper tools (grim, gnome-screenshot, ydotool) so Detect can warn when
// a Wayland session lacks the tooling that makes capture/input actually work.
func hasCmd(name string) bool {
	_, err := exec.LookPath(name)
	return err == nil
}

// Capabilities describes what the computer-use server can do on this platform.
type Capabilities struct {
	HasDisplay           bool     `json:"has_display"`
	Platform             string   `json:"platform"`
	DisplayServer        *string  `json:"display_server,omitempty"`
	DisplayResolution    *[2]int  `json:"display_resolution,omitempty"`
	ScaleFactor          float64  `json:"scale_factor"`
	AccessibilityGranted bool     `json:"accessibility_granted"`
	IsAdmin              bool     `json:"is_admin"`
	Issues               []string `json:"issues"`
}

// Detect returns the current platform's capabilities.
func Detect() Capabilities {
	platform := runtime.GOOS
	hasDisplay, displayServer := detectDisplayServer()

	var displayResolution *[2]int
	if hasDisplay {
		if res := detectResolution(); res != nil {
			displayResolution = res
		}
	}

	scaleFactor := 1.0
	if displayResolution != nil {
		w := displayResolution[0]
		if w > 2000 {
			scaleFactor = float64(w) / 1400.0
		}
	}

	accessibilityGranted := checkAccessibilityPermission()
	isAdmin := checkAdminStatus()

	var issues []string
	if !hasDisplay {
		issues = append(issues, "No display server detected. Computer use tools unavailable.")
	}
	if hasDisplay && !accessibilityGranted && platform == "darwin" {
		issues = append(issues, "Accessibility permission required. Grant in: System Preferences → Privacy & Security → Accessibility → Enable AgentBoster")
	}
	// On a native Wayland session, pixel capture (XGetImage) and synthetic input
	// (XTest) only work for Xwayland-backed windows — native Wayland windows show
	// as black frames and ignore the fake events. The Wayland-native helpers close
	// that gap, so flag them as actionable issues when missing rather than letting
	// screenshots silently come back black.
	if hasDisplay && displayServer != nil && *displayServer == "wayland" {
		if !hasCmd("grim") && !hasCmd("gnome-screenshot") {
			issues = append(issues, "Wayland session without a screenshot helper: install 'grim' (wlroots: sway/Hyprland) or 'gnome-screenshot' (GNOME), or native-window captures will be black.")
		}
		if !hasCmd("ydotool") {
			issues = append(issues, "Wayland session without 'ydotool': mouse/keyboard input can only reach Xwayland windows. Install ydotool and run ydotoold (needs /dev/uinput) for native Wayland input.")
		}
	}

	return Capabilities{
		HasDisplay:           hasDisplay,
		Platform:             platform,
		DisplayServer:        displayServer,
		DisplayResolution:    displayResolution,
		ScaleFactor:          scaleFactor,
		AccessibilityGranted: accessibilityGranted,
		IsAdmin:              isAdmin,
		Issues:               issues,
	}
}

func detectDisplayServer() (bool, *string) {
	switch runtime.GOOS {
	case "darwin":
		s := "quartz"
		return true, &s
	case "windows":
		s := "win32"
		return true, &s
	case "linux":
		if os.Getenv("WAYLAND_DISPLAY") != "" {
			s := "wayland"
			return true, &s
		}
		if os.Getenv("DISPLAY") != "" {
			s := "x11"
			return true, &s
		}
		return false, nil
	default:
		return false, nil
	}
}

func detectResolution() *[2]int {
	displays, err := screenshot.GetDisplays()
	if err != nil || len(displays) == 0 {
		return nil
	}

	bounds := displays[0].Bounds
	w := bounds.Dx()
	h := bounds.Dy()

	return &[2]int{w, h}
}
